index.js

```diff
@@ -0,0 +1,303 @@
+//
+Computation.prototype._mapMetadata = function (s) {
+  return typeof s === 'string'
+    ? new StreamMetadata({ name: s, grouping: 'SHUFFLE' })
+    : new StreamMetadata({ name: s.name, grouping: groupMap(s.grouping) || 'SHUFFLE' });
+};
+
+//
+// Connect to the client proxy
+//
+Computation.prototype.connect = function (host, port, callback) {
+  var self = this;
+  //
+  // Set the values and initialize the proxy
+  //
+  this.host = host;
+  this.port = port;
+  this.proxy = this.createProxy(host, port);
+
+  this.metadata.proxyEndpoint = new Endpoint({ ip: host, port: port });
+
+  //
+  // Alias functions over so they are properly called by the thrift machinery
+  //
+  this.getState = this.proxy.getState;
+  this.setState = this.proxy.setState;
+
+  //
+  // Check if there is a result we care about
+  //
+  this.proxy.registerWithScheduler(this.metadata, function (err) {
+    if (err) {
+      return callback
+        ? callback(err)
+        : self.emit('error', err);
+    }
+
+    self.emit('register', self.metadata);
+
+    if (callback) callback();
+  });
+
+};
+
+//
+// Create a bolt service proxy through instantiating a proper thrift connection
+//
+Computation.prototype.createProxy = function (host, port) {
+  this.connection = thrift.createConnection(host, port, {
+    transport: FramedTransport(),
+    protocol: BinaryProtocol()
+  });
+
+  this.connection.on('error', this.emit.bind(this, 'error'));
+  this.connection.on('connect', this.emit.bind(this, 'connection'));
+  //
+  // Return a BoltProxyService
+  //
+  return thrift.createClient(BoltProxyService, this.connection);
+};
+
+//
+// A wrapper around the thrift ComputationTx to initialize values
+//
+function Transaction() {
+  return new Tx({
+    records: [],
+    timers: {}
+  });
+}
+//
+// A context that wraps the transaction and provides methods to modify each
+// transaction sent to concord
+//
+function Context(tx) {
+  this.__tx = tx;
+}
+
+//
+// Accepts the stream name, the key and the value
+//
+Context.prototype.produceRecord = function (stream, key, value) {
+  this.__tx.records.push(new Record({
+    key: key,
+    value: value,
+    userStream: stream
+  }));
+};
+
+Context.prototype.setTimer = function (key, time) {
+  this.__tx.timers[key] = time;
+};
+
+//
+// Handle mapping a grouping name to the proper TYPE, otherwise return an error
+//
+function groupMap(grouping) {
+  return Types.streamGrouping[snakeCase(grouping).toUpperCase()]
+}
+
+//
+// Naive snakeCase implementation
+//
+function snakeCase(text) {
+  return text.replace(/[a-z][A-Z]/g, function (ch) {
+    return ch[0] + '_' + ch[1].toLowerCase();
+  });
+};
+
+//
+// Parse environment variables and create a thrift service based on the
+// computation
+//
+exports.run = function run(computation) {
+  var server = thift.createServer(ComputationService, computation);
+
+  //
+  // Send a start event when we are fully connected to everything
+  //
+  var finish = after(2, function () {
```

Comment:
@bigs it basically waits for 2 function calls to happen before emitting. So its a simple counter behind the scenes before it executes the given callback.